File: examples/structdunders.py
class Number:
    def __init__(self, x):
        self.x = x

    def __add__(self, other: 'Number'):
        return Number(self.x + other.x)

    def __mul__(self, other: 'Number'):
        return Number(self.x * other.x)

    def __str__(self):
        return str(self.x)


class ComplexNumber:
    def __init__(self, real, imag):
        self.real = real
        self.imag = imag

#    def __add__(self, other: 'ComplexNumber'):
#        return ComplexNumber(self.real + other.real, self.imag + other.imag)

    def __str__(self):
        return str(self.real) + "+" + str(self.imag) + "i"


def main():
    five = Number(5)
    six = Number(6)
    eleven = five + six
    print(eleven)
    fifty_five = eleven * five
    print(fifty_five)
    i = ComplexNumber(Number(0), Number(-1))
    n = ComplexNumber(Number(3), Number(-6))
    # print(i + n)
    print(i, n)


if __name__ == '__main__':
    main()

File: examples/structdunders.go
package main

import "fmt"

type Number struct {
	x int
}

func NewNumber(x int) (self *Number) {
	self = new(Number)
	self.x = x
	return
}

func (self *Number) Add(other Number) *Number {
	return NewNumber(self.x + other.x)
}

func (self *Number) Mul(other Number) *Number {
	return NewNumber(self.x * other.x)
}

func (self *Number) String() string {
	return fmt.Sprintf("%v", self.x)
}

type ComplexNumber struct {
	real *Number
	imag *Number
}

func NewComplexNumber(real *Number, imag *Number) (self *ComplexNumber) {
	self = new(ComplexNumber)
	self.real = real
	self.imag = imag
	return
}

func (self *ComplexNumber) String() string {
	return fmt.Sprintf("%v", self.real) + "+" + fmt.Sprintf("%v", self.imag) + "i"
}

func main() {
	five := NewNumber(5)
	six := NewNumber(6)
	eleven := five.Add(*six)
	fmt.Println(eleven)
	fifty_five := five.Mul(*eleven)
	fmt.Println(fifty_five)
	i := NewComplexNumber(NewNumber(0), NewNumber(-1))
	n := NewComplexNumber(NewNumber(3), NewNumber(-6))
	fmt.Println(i, n)
}
